hivemq-edge/src/main/java/com/hivemq/persistence/domain/DomainTagAddResult.java

```diff
@@ -18,41 +18,54 @@
 import com.hivemq.extension.sdk.api.annotations.NotNull;
 import com.hivemq.extension.sdk.api.annotations.Nullable;
 
+import java.util.Optional;
+
 public class DomainTagAddResult {
 
     private final @NotNull DomainTagPutStatus dataPolicyPutStatus;
-    private final @Nullable String errorMessage;
+    private final @Nullable Optional<String> errorMessage;
+    private final @NotNull Optional<String> adapterIdOfOwningAdapter;
 
     public DomainTagAddResult(
-            final @NotNull DomainTagPutStatus dataPolicyPutStatus, final @Nullable String errorMessage) {
+            final @NotNull DomainTagPutStatus dataPolicyPutStatus, final @Nullable String adapterIdOfOwningAdapter, final @Nullable String errorMessage) {
         this.dataPolicyPutStatus = dataPolicyPutStatus;
-        this.errorMessage = errorMessage;
+        this.errorMessage = Optional.ofNullable(errorMessage);
+        this.adapterIdOfOwningAdapter = Optional.ofNullable(adapterIdOfOwningAdapter);
     }
 
     public static @NotNull DomainTagAddResult success() {
-        return new DomainTagAddResult(DomainTagPutStatus.SUCCESS, null);
+        return new DomainTagAddResult(DomainTagPutStatus.SUCCESS, null,null);
     }
 
-    public static @NotNull DomainTagAddResult failed(final @NotNull DomainTagPutStatus putStatus) {
-        return new DomainTagAddResult(putStatus, null);
+    public static @NotNull DomainTagAddResult failed(
+            final @NotNull DomainTagPutStatus putStatus,
+            final @NotNull String adapterIdOfOwningAdapter) {
+        return new DomainTagAddResult(putStatus, adapterIdOfOwningAdapter, null);
     }
 
     public static @NotNull DomainTagAddResult failed(
-            final @NotNull DomainTagPutStatus putStatus, final @Nullable String errorMessage) {
-        return new DomainTagAddResult(putStatus, errorMessage);
+            final @NotNull DomainTagPutStatus putStatus,
+            final @Nullable String adapterIdOfOwningAdapter,
+            final @Nullable String errorMessage) {
+        return new DomainTagAddResult(putStatus, adapterIdOfOwningAdapter, errorMessage);
     }
 
     public @NotNull DomainTagPutStatus getDomainTagPutStatus() {
         return dataPolicyPutStatus;
     }
 
-    public @Nullable String getErrorMessage() {
+    public @Nullable Optional<String> getErrorMessage() {
         return errorMessage;
     }
 
+    public @NotNull Optional<String> getAdapterIdOfOwningAdapter() {
```

Comment:
why optional? Shouldnt we always have an adapterId?

Reply:
Not if the add failed.
DomainTagAddResult has the states success and failed, each one having different fields filled.